TSPL.docc/LanguageGuide/Concurrency.md

```diff
@@ -564,33 +637,192 @@ the explicit parent-child relationships between tasks
+
+```
+let photos = await withTaskGroup(of: Data.self) { taskGroup in
     let photoNames = await listPhotos(inGallery: "Summer Vacation")
     for name in photoNames {
         taskGroup.addTask { await downloadPhoto(named: name) }
     }
+
+    var results: [Data] = []
+    for await photo in taskGroup {
+        results.append(photo)
+    }
+    return results
 }
 ```
 
-<!--
-  TODO walk through the example
+Like the previous example,
+this example creates a child task for each photo to download it.
+Unlike the previous example,
+the `for`-`await`-`in` loop waits for the next child task to finish,
+appends the result of that task to the array of results,
+and then continues waiting until all child tasks have finished.
+Finally,
+the task group returns the array of downloaded photos
+as its overall result.
+
+
+### Task Cancellation
+
+Swift concurrency uses a cooperative cancellation model.
+Each task checks whether it has been canceled
+at the appropriate points in its execution,
+and responds to cancellation an appropriate way.
+Depending on the work you're doing,
+that usually means one of the following:
+
+- Throwing an error like `CancellationError`
+- Returning `nil` or an empty collection
+- Returning the partially completed work
+
+Downloading pictures could take a long time
+if the pictures are large or the network is slow.
+To let the user stop this work,
+without waiting for all of the tasks to complete,
+the tasks need check for cancellation and stop running if they are canceled.
+There are two ways a task can do this:
+by calling the [`Task.checkCancellation()`][] method,
+or by reading the [`Task.isCancelled`][] property.
+Calling `checkCancellation()` throws an error if the task is canceled;
+a throwing task can let that error propagate out of the task,
+stopping all of the task's work.
+This has the advantage of being simple to implement and understand.
+For more flexibility, use the `isCancelled` property,
+which lets you perform clean-up work as part of stopping the task,
+like closing network connections and deleting temporary files.
+
+[`Task.checkCancellation()`]: https://developer.apple.com/documentation/swift/task/3814826-checkcancellation
+[`Task.isCancelled`]: https://developer.apple.com/documentation/swift/task/3814832-iscancelled
+
+```
+let photos = await withTaskGroup(of: Data.self) { taskGroup in
+    let photoNames = await listPhotos(inGallery: "Summer Vacation")
+    for name in photoNames {
+        taskGroup.addTaskUnlessCancelled {
+            guard isCancelled == false else { return nil }
+            return await downloadPhoto(named: name)
+        }
+    }
+
+    var results: [Data] = []
+    for await photo in taskGroup {
+        if let photo { results.append(photo) }
+    }
+    return results
+}
+```
+
+<!-- XXX TR:
+Should the listing above add a call to taskGroup.cancelAll()
+to show that part of cancellation?
+If so, what's the best practice for where to put that?
```

Comment:
Probably not necessary here.

We could have another example where we look at the values returned and one of them is like "time to give up" 

```
for await value in group { 
  switch value {
  case .apiLimitExceeded: group.cancelAll() // we won't be able to get more results anyway, let's cancel
  case .data(data): ... 
  }
}
```

?

Reply:
This might be a good addition to TSPL or to the `TaskGroup` reference.  I'll mark this specific PR comment as "resolved" so it doesn't block merging.